torch/_inductor/fx_passes/b2b_gemm.py

```diff
@@ -0,0 +1,160 @@
+    PatternMatcherPass,
+    register_graph_pattern,
+)
+from ..select_algorithm import (
+    autotune_select_algorithm,
+    TritonTemplate,
+    TritonTemplateCaller,
+)
+from ..utils import ceildiv
+
+aten = torch.ops.aten
+
+
+def b2b_gemm_grid(M, P, meta):
+    return (ceildiv(M, meta["BLOCK_SIZE_M"]) * ceildiv(P, meta["BLOCK_SIZE_P"]), 1, 1)
+
+b2b_gemm_template = TritonTemplate(
+    name="b2b_gemm",
+    grid=b2b_gemm_grid,
+    debug=False,
+    source=r"""
+{{def_kernel("A", "B", "C")}}
+
+
+    # B2B_GEMM_TRITON_ENTRANCE
+    # TODO: handle the non-divisible case
+    M = {{size("A", 0)}}
+    N = {{size("A", 1)}}
+    O = {{size("C", 0)}}
+    P = {{size("C", 1)}}
+
+    stride_am = {{stride("A", 0)}}
+    stride_an = {{stride("A", 1)}}
+    stride_bn = {{stride("B", 0)}}
+    stride_bo = {{stride("B", 1)}}
+    stride_co = {{stride("C", 0)}}
+    stride_cp = {{stride("C", 1)}}
+
+    # output (M * P) block ids
+    pid = tl.program_id(axis=0)
+    m_block_id = pid // (P // BLOCK_SIZE_P)
+    p_block_id = pid % (P // BLOCK_SIZE_P)
+
+    # internal block numbers
+    num_n_block = tl.cdiv(N, BLOCK_SIZE_N)
+    num_o_block = tl.cdiv(O, BLOCK_SIZE_O)
+
+    # accumulator
+    acc = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_P), dtype=tl.float32)
+
+    for n_block_id in range(num_n_block):
+        a_ptrs = A + (
+            (m_block_id * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M))[:, None] * stride_am +
+            (n_block_id * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N))[None, :] * stride_an
+        )
+        a = tl.load(a_ptrs)
+        for o_block_id in range(num_o_block):
+            b_ptrs = B + (
+                (n_block_id * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N))[:, None] * stride_bn +
+                (o_block_id * BLOCK_SIZE_O + tl.arange(0, BLOCK_SIZE_O))[None, :] * stride_bo
+            )
+            b = tl.load(b_ptrs)
+            c_ptrs = C + (
+                (o_block_id * BLOCK_SIZE_O + tl.arange(0, BLOCK_SIZE_O))[:, None] * stride_co +
+                (p_block_id * BLOCK_SIZE_P + tl.arange(0, BLOCK_SIZE_P))[None, :] * stride_cp
+            )
+            c = tl.load(c_ptrs)
+            acc += tl.dot(tl.dot(a, b, out_dtype=tl.float16), c, out_dtype=tl.float16)
+
+    # store
+    idx_m = (m_block_id * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M))[:, None]
+    idx_p = (p_block_id * BLOCK_SIZE_P + tl.arange(0, BLOCK_SIZE_P))[None, :]
+    mask = (idx_m < M) & (idx_p < P)
+
+    {{store_output(("idx_m", "idx_p"), "acc", "mask")}}
+""",
+)
+
+
+B2B_GEMM_PASS = PatternMatcherPass(
+    prevent_match_across_mutations=True,
+    pass_name="b2b_gemm_pass",
+)
+
+
+def can_apply_b2b_gemm(match: Match) -> bool:
+    if not all(["val" in arg.meta for arg in match.args]):
+        return False
+    mats = [arg.meta["val"] for arg in match.args]
+    if not all([mat.is_cuda for mat in mats]):
+        return False
+    if not all([len(mat.shape) == 2 for mat in mats]):
+        return False
+    mat1, mat2, mat3 = mats
```

Comment:
nit: again I don't feel strongly because it's just stylistic, but if you actually still need to split out mat1, mat2, mat3 here, I think I'd just go back to your previous style. Or use mats[0], mats[1], etc. Again, up to you.